src/pybind/mgr/dashboard/controllers/auth.py

```diff
@@ -28,33 +29,48 @@ class Auth(RESTController):
     """
     Provide authenticates and returns JWT token.
     """
-
     def create(self, username, password):
         user_data = AuthManager.authenticate(username, password)
         user_perms, pwd_expiration_date, pwd_update_required = None, None, None
-        if user_data:
-            user_perms = user_data.get('permissions')
-            pwd_expiration_date = user_data.get('pwdExpirationDate', None)
-            pwd_update_required = user_data.get('pwdUpdateRequired', False)
-
-        if user_perms is not None:
-            logger.debug('Login successful')
-            token = JwtManager.gen_token(username)
-            token = token.decode('utf-8')
-            cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
-            return {
-                'token': token,
-                'username': username,
-                'permissions': user_perms,
-                'pwdExpirationDate': pwd_expiration_date,
-                'sso': mgr.SSO_DB.protocol == 'saml2',
-                'pwdUpdateRequired': pwd_update_required
-            }
+        max_attempt = Settings.ACCOUNT_LOCKOUT_ATTEMPTS
+        user_attempts = mgr.ACCESS_CTRL_DB.get_attempt(username)
+        if max_attempt == 0 or user_attempts < max_attempt:
+            if user_data:
+                user_perms = user_data.get('permissions')
+                pwd_expiration_date = user_data.get('pwdExpirationDate', None)
+                pwd_update_required = user_data.get('pwdUpdateRequired', False)
 
-        logger.debug('Login failed')
-        raise DashboardException(msg='Invalid credentials',
-                                 code='invalid_credentials',
-                                 component='auth')
+            if user_perms is not None:
+                logger.info('Login successful: %s', username)
+                mgr.ACCESS_CTRL_DB.update_attempt(username, 0)
+                mgr.ACCESS_CTRL_DB.save()
+                token = JwtManager.gen_token(username)
+                token = token.decode('utf-8')
+                cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
+                return {
+                    'token': token,
+                    'username': username,
+                    'permissions': user_perms,
+                    'pwdExpirationDate': pwd_expiration_date,
+                    'sso': mgr.SSO_DB.protocol == 'saml2',
+                    'pwdUpdateRequired': pwd_update_required
+                }
+            user_attempts += 1
+            mgr.ACCESS_CTRL_DB.update_attempt(username, user_attempts)
+            mgr.ACCESS_CTRL_DB.save()
```

Comment:
```suggestion
            with mgr.ACCESS_CTRL_DB.lock:
                mgr.ACCESS_CTRL_DB.update_attempt(username, mgr.ACCESS_CTRL_DB.get_attempt(username) + 1)
                mgr.ACCESS_CTRL_DB.save()
```

Reply:
I went with the other way and applied the lock inside the `increment_attempt()` method